aten/src/ATen/native/cpu/FusedAdamKernel.cpp

```diff
@@ -0,0 +1,147 @@
+#include <ATen/core/Tensor.h>
+#include <ATen/Parallel.h>
+#include <ATen/native/DispatchStub.h>
+#include <ATen/native/fused_adam.h>
+#include <ATen/Dispatch.h>
+#include <ATen/cpu/vec/vec.h>
+namespace at::native {
+
+namespace{
+
+template <typename scalar_t>
+void adam_fused_step_impl(
+    const at::Tensor& param,
+    const at::Tensor& grad,
+    const at::Tensor& exp_avg,
+    const at::Tensor& exp_avg_sq,
+    const at::Tensor& max_exp_avg_sq,
+    const at::Tensor& state_step,
+    const double lr,
+    const double beta1,
+    const double beta2,
+    const double weight_decay,
+    const double eps,
+    const bool amsgrad,
+    const bool maximize,
+    const float* grad_scale_ptr) {
+  int64_t step = state_step.item<int64_t>();
+  scalar_t* param_data = param.data_ptr<scalar_t>();
+  scalar_t* exp_avg_data = exp_avg.data_ptr<scalar_t>();
+  scalar_t* exp_avg_sq_data = exp_avg_sq.data_ptr<scalar_t>();
+  scalar_t* max_exp_avg_sq_data = amsgrad ? max_exp_avg_sq.data_ptr<scalar_t>() : nullptr;
+  scalar_t* grad_data = grad.data_ptr<scalar_t>();
+
+  // need to use double here to align with non-fused adam
+  double bias_correction1 = 1 - std::pow(beta1, step);
+  double step_size =  lr / bias_correction1;
+  double bias_correction2 = 1 - std::pow(beta2, step);
+  double exp_avg_grad_coefficient = 1 - beta1;
+  double exp_avg_sq_grad_coefficient = 1 - beta2;
+  double bias_correction2_sqrt = std::sqrt(bias_correction2);
+
+  using Vec = at::vec::Vectorized<scalar_t>;
```

Comment:
All the arithmetic operations are applied on the `Vec` directly. Is it correct for bfloat16 and half? Shall we use fp32 as intermediate data type instead? It has high precision and also better performance.

Reply:
Sure, I have considered it. And I will enhance it for `fp16/bf16`.

Previously I just think `float32` is most import because for training with `autocast`, the `param,grad,exp_avg,etc` here are always `float32`.
Only when user explicitly call `model.to(torch.bfloat16)` will hit the case you mentioned and I think this may not be our proposed usage.

Reply:
Updated as https://github.com/pytorch/pytorch/pull/123074/commits/fe2a9e6ce594f13cf2e6c8225f156930d347558a#diff-ca89133ed5592a5cd9a1fb8ce961b28b5842b3380dc90ecde8c3a8fd1cf21e3e.